scraper/newsscrapper/spiders/cnn.py

```diff
@@ -3,32 +3,53 @@
 from scrapy.linkextractors import LinkExtractor
 from scrapy.item import Item, Field
 from newsscrapper.items import Article
+from scrapy.loader.processors import TakeFirst, MapCompose
 
+#   to install pytz module for converting time to UTC
+#   pip install pytz
 #
 #   cd to spiders
 #   scrapy crawl cnn (to run spider)
-#   scrapy shell "https://edition.cnn.com/"
+#   scrapy shell "https://edition.cnn.com/world"
 #
 
 class CNNSpider(scrapy.Spider):
     name = "cnn"
     allowed_domains = ['edition.cnn.com']
-    start_urls = ['https://edition.cnn.com/']
+    start_urls = ['https://edition.cnn.com/world']
+    output_file = 'data/cnn_output.json'
 
     def parse(self, response):
-        for href in response.xpath('//a[contains(@class, "gs-c-promo-heading")]/@href'): 
-            print(href)
+        for href in response.xpath('//div[@class="stack"]//a/@href'):
             url = response.urljoin(href.extract())
-            
             yield scrapy.Request(url, callback = self.getArticle)
+        
     
     def getArticle(self, response):
         item = Article()
-        item['name'] = response.xpath('//h1/text()').get()
-        item['author'] = response.xpath('//div[contains(@class, "ssrcss-68pt20-Text-TextContributorName")]/text()').get()
+        
+        item['name'] = response.xpath('//h1[@id="maincontent"]/text()').get().strip()
+
+        item['author'] = response.xpath('//span[@class="byline__name"]/text()').get()
+
+        item['publication_date'] = response.xpath('//div[@class="timestamp"]/text()').get()
+        item['publication_date'] = item['publication_date'].split(',', 1)[1].strip() if item['publication_date'] else "Not Found"
+                
+        body_text = [text.strip() for text in response.xpath('//div[@class="article__content"]/p//text()').getall()]
+        body_text = ' '.join(body_text)
+        item['body'] = body_text
+
+        item['category'] = 'news'
+
         item['source_url'] = response.url
+        item['cover_url'] = response.xpath('//picture[@class="image__picture"]/source/@srcset').get()
 
-        yield item
-        
+        if response.xpath('//div[@class="gallery-inline__container"]'):
+            item['body'] = response.xpath('//span[@class="inline-placeholder"]/text()').getall()
+            # item['cover_url'] = response.xpath('//picture[@class="image_gallery-image__picture"]/img/@src').getall()
 
+        for key in ['name', 'author', 'publication_date', 'cover_url']:
+            if not item[key]:
+                item[key] = "Not Found"
```

Comment:
Storing a null value as a string of "Not Found" makes searching more difficult, and technically breaks first normal form ( each field stores once piece of information )

Python uses [*None*](https://www.w3schools.com/python/ref_keyword_none.asp) to communicate null values, this will probably be converted to either `null` or `undefined` when parsed to JSON.

The database has constraints to protect data, not encourage loopholes. If some articles lack these fields and it cant be determined, we will adjust the schema.

Reply:
@AardhynLavender, I agree, it is not the best thing to do, but some articles do not have information about author, name or date. I am setting them to "Not Found" because of the rules for those fields that don't let them to be `null` (can see in database/src/init.sql). It was throwing me the error and it would not send the data to db when I kept those fields empty. I can just not include articles that do not have author, name or date into the data that I am sending to db, if you don't want to use this method. But I would not want to lose data for the article that doesn't have a value just for one field.

Reply:
If some articles have no author, we will make the field nullable.

If its impossible to tell what the publication date is for an article, we will make it nullable. ( I'm thinking a `retrieved_date` field that's automatically set might be the solution for this one ).

lets not set anything to "not found". let me know the fields that might be null.

Reply:
@AardhynLavender. The fields that appear to be `null` frequently are: `name`, `author`, `publication_date`

Reply:
@ArtemKech, I'll make `publication_date` and `name` nullable in the main branch. 

Maybe set `author` to the website being scraped if you cant find a person.

Reply:
Do some articles really have no name?

Reply:
@AardhynLavender I believe there was one layout from **_BBC_** that had multiple pictures and bodies for every one of them. It used the first body's text in the list as a title for the article, what means there is technically no article name. I will show you exactly what I have described if I find this type of article.